test/Microsoft.DotNet.ShellShimMaker.Tests/ShellShimMakerTests.cs

```diff
@@ -0,0 +1,153 @@
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.InteropServices;
+using FluentAssertions;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.Tools.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.DotNet.ShellShimMaker.Tests
+{
+    public class ShellShimMakerTests : TestBase
+    {
+        [Fact]
+        public void GivenAnExecutablePathItCanGenerateShimFile()
+        {
+            var outputDll = MakeHelloWorldExecutableDll();
+
+            var muxer = new Muxer();
+            var shellShimMaker = new ShellShimMaker(Path.GetDirectoryName(muxer.MuxerPath));
+            var shellCommandName = nameof(ShellShimMakerTests) + Path.GetRandomFileName();
+
+            shellShimMaker.CreateShim(
+                outputDll.FullName,
+                shellCommandName);
+            var stdOut = ExecuteInShell(shellCommandName);
+
+            stdOut.Should().Contain("Hello World");
+
+            // Tear down
+            shellShimMaker.Remove(shellCommandName);
+        }
+
+        [Fact]
+        public void GivenAnExecutablePathWithExistingSameNameShimItThrows()
+        {
+            var muxer = new Muxer();
+            var pathToPlaceShim = Path.GetDirectoryName(muxer.MuxerPath);
+            var shellCommandName = nameof(ShellShimMakerTests) + Path.GetRandomFileName();
+
+            MakeNameConflictingCommand(pathToPlaceShim, shellCommandName);
+
+            var shellShimMaker = new ShellShimMaker(pathToPlaceShim);
+
+            Action a = () => shellShimMaker.EnsureCommandNameUniqueness(shellCommandName);
+            a.ShouldThrow<GracefulException>()
+                .And.Message
+                .Should().Contain($"Failed to create tool {shellCommandName}, a command with the same name existed");
+
+            // Tear down
+            shellShimMaker.Remove(shellCommandName);
+        }
+
+
+        [Fact]
+        public void GivenAnExecutablePathWithoutExistingSameNameShimItShouldNotThrow()
+        {
+            var muxer = new Muxer();
+            var pathToPlaceShim = Path.GetDirectoryName(muxer.MuxerPath);
+            var shellCommandName = nameof(ShellShimMakerTests) + Path.GetRandomFileName();
+
+            var shellShimMaker = new ShellShimMaker(pathToPlaceShim);
+
+            Action a = () => shellShimMaker.EnsureCommandNameUniqueness(shellCommandName);
+            a.ShouldNotThrow();
+
+            // Tear down
+            shellShimMaker.Remove(shellCommandName);
+        }
+
+        private static void MakeNameConflictingCommand(string pathToPlaceShim, string shellCommandName)
+        {
+            File.WriteAllText(Path.Combine(pathToPlaceShim, shellCommandName), string.Empty);
+        }
+
+
+        private static string ExecuteInShell(string shellCommandName)
+        {
+            string stdOut;
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                ExecuteAndCaptureOutputWithAssert(new ProcessStartInfo
```

Comment:
We have code that will start processes for you and capture output. Why did you need to write your own with ExecuteAndCaptureOutput?

Reply:
I am aware of it, however it is internal in util project. I was hesitate to use it with internalVisibleTo, since most util project is considered internal already